### ISSUE TITLE
Expected tokens optimized out in favor of children

### ISSUE THREAD

**Issue by Synival** [orangeduck/mpc#54]
I'm not sure if this is correct behavior or not, but it's causing me some execution issues. 

My rule <resource> expects `<inner_block>`, which contains `(<statement> | <comment>)*`. In the case that there is only 1 child in <inner_block>, the <inner_block> is optimized out in favor of the child, which I do not expect to interpret at the <resource> level.  Here's the input:

```
[my_func]
   echo (a b c)
```

...and the grammar as defined by mpca_lang():

```
qscript        : /^/ (<comment> | <resource>)* /$/ ;
   comment     : '#' /[^\\n]*/ ;
resource       : '[' (<rtype> <rname>) ']' <inner_block> ;
   rtype       : /[*]*/ ;
   rname       : <qstring> ;

inner_block    : (<comment> | <statement>)* ;
   statement   : <function> '(' (<comment> | <parameter> | <block>)* ')'  <seperator> ;
   function    : <qstring> ;
   parameter   : (<statement> | <literal>) ;
      literal  : (<number> | <qstring>) <seperator> ;
   block       : '{' <inner_block> '}' ;
   seperator   : ',' | \"\" ;

qstring        : (<complexstr> | <simplestr>) <qstring>* ;
   simplestr   : /[a-zA-Z0-9_!@#$%^&\\*_+\\-\\.=\\/<>]+/ ;
   complexstr  : (/\"[^\"]*\"/ | /'[^']*'/) ;

number         : (<float> | <int>) ;
   float       : /[-+]?[0-9]+\\.[0-9]+/ ;
   int         : /[-+]?[0-9]+/ ;
```

Here are the results from mpc_ast_print():

```
> 
  regex 
  resource|> 
    char:1:1 '['
    rtype|regex 
    rname|qstring|simplestr|regex:1:2 'my_func'
    char:1:9 ']'
    statement|>     <--------------- shouldn't expect a statement in <resource>!
      function|qstring|simplestr|regex:2:4 'echo'
      char:2:9 '('
      literal|> 
        qstring|> 
          simplestr|regex:2:10 'a'
          qstring|> 
            simplestr|regex:2:12 'b'
            qstring|simplestr|regex:2:14 'c'
        seperator|string 
      char:2:15 ')'
      seperator|string 
  regex
```

I've noticed that when I comment out a certain optimization at mpc.c +2695:  
`  if (a->children_num == 1) { return a; }`  

...I get the desired results:
```
(snip)
    char:1:9 ']'
    inner_block|> 
      statement|> 
        function|qstring|simplestr|regex:2:4 'echo'
(etc)
```

I've found that changing the <inner_block> to:  
`inner_block    : (<comment> | <statement>)* \"\";`  

...which forces a <string> token after <inner_block>, which is a nice work-around, but doesn't feel like a solution.

**Edit**: Otherwise awesome library, btw :)
**Edit 2**: Formatting

**Comment by Synival** [comment]
Actually, when optimized, should the type be collapsed to `inner_block|statement|>`, just as it is for `rname|qstring|simplestr|regex`?

**Comment by orangeduck** [comment]
Yeah the tag should be collapsing to `inner_block|statement|>`. I wonder if this is to do with the optimisations I added in. At the bottom of `mpc.c` you could try changing

```c
void mpc_optimise(mpc_parser_t *p) {
  mpc_optimise_unretained(p, 1);
}
```
to
```
void mpc_optimise(mpc_parser_t *p) { return; }
```

and see what happens? I'll investigate this in a bit more detail a the weekend.

**Comment by orangeduck** [comment]
Hey,

Thanks - I found the bug. It was nothing to do with the optimisations - instead it was some bug with how single child asts are merged. This should be fixed in 2a19e4c45161f99dee7f1a378b3e4d11e2af7162. I hope this wont break any existing code but I think it will be okay in most cases as it shouldn't change the tree structure output by the parser, just add some additional tags.

- Dan
